### PR TITLE
    Adds an interactive logs subcommand for pipeline

### PR DESCRIPTION
**Pull request by pradeepitm12**
This adds an interactive logs subcommand for pipeline to show logs
    of the pipeline.
    
    There are three different ways to run the command
    
        - tkn pipeline logs
          this is an interactive mode where it asks for pipeline,
          pipelinerun and then show logs
        - tkn pipeline logs <pipeline_name>
          this is also an interactive mode where it just ask for
          pipelinerun name and then show logs
        - tkn pipeline logs <pipeline_name> --last
          this shows logs of the pipeline for last run
    
    Signed-off-by: Pradeep Kumar <pradkuma@redhat.com>

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```
release-note
```

### CODE REVIEW

File: pkg/cmd/pipeline/logs.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/tektoncd/cli/pkg/cmd/pipelinerun"
+	"github.com/tektoncd/cli/pkg/helper/pods"
+	"github.com/tektoncd/cli/pkg/helper/pods/stream"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// LogOptions provides options on what logs to fetch. An empty LogOptions
+// implies fetching all logs including init steps
+type LogOptions struct {
```
Comment (hrishin) [inline pkg/cmd/pipeline/logs.go]: can we re-use it from `pipleinerun` package instead of defining it again here?

File: pkg/cmd/pipeline/logs.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	c.Flags().BoolVarP(&opts.follow, "follow", "f", false, "stream live logs")
+	c.Flags().StringVar(&opts.pipelinerunName, "pipelinerun", "", "show logs for mentioned pipelinerun only")
+	return c
+}
+
+func (lo *LogOptions) run() error {
+	if lo.pipelineName == "" {
+		return fmt.Errorf("missing mandatory argument pipeline")
+	}
+	fmt.Println("Pradeep New : ", lo.pipelinerunName)
```
Comment (hrishin) [inline pkg/cmd/pipeline/logs.go]: Nit: needed it here?

File: pkg/cmd/pipeline/logs.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		Args:    cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.pipelineName = args[0]
+			opts.stream = &cli.Stream{
+				Out: cmd.OutOrStdout(),
+				Err: cmd.OutOrStderr(),
+			}
+			opts.streamer = pods.NewStream
+
+			return opts.run()
```
Comment (hrishin) [inline pkg/cmd/pipeline/logs.go]: we can just query the last `pipelinerun` then pass it to `pipelinerun.LogOption`  then invoke Run()?

Comment (pradeepitm12) [reply pkg/cmd/pipeline/logs.go]: Make sense! Done

File: pkg/cmd/pipeline/logs.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  # show the logs of the latest pipelinerun of given pipeline
+	tkn pipeline logs foo -n bar
+
+   `
+	c := &cobra.Command{
+		Use: "logs",
+		DisableFlagsInUseLine: true,
+		Short:   "Show the logs of latest pipelinerun of given pipeline ",
+		Example: eg,
+		Args:    cobra.ExactArgs(1),
```
Comment (sthaha) [inline pkg/cmd/pipeline/logs.go]: Please consider using go-fmt

File: pkg/cmd/pipeline/logs.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+func logCommand(p cli.Params) *cobra.Command {
+	opts := &pipelinerun.LogOptions{Params: p}
+	eg := `
+  # show the logs of the latest pipelinerun of given pipeline
+	tkn pipeline logs foo -n bar
+
+   `
+	c := &cobra.Command{
+		Use: "logs",
```
Comment (sthaha) [inline pkg/cmd/pipeline/logs.go]: @vdemeester do you know if we can add a linter to our ci check?

File: pkg/cmd/pipeline/logs.go (hunk 4 above)
Comment (sthaha) [reply pkg/cmd/pipeline/logs.go]: Please use `NameArg` that is part of `start` command, `perhaps move the validations to `validation` package

File: pkg/cmd/pipeline/logs.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  # show the logs of the latest pipelinerun of given pipeline
+	tkn pipeline logs foo -n bar
+
+   `
+	c := &cobra.Command{
+		Use:                   "logs",
+		DisableFlagsInUseLine: true,
+		Short:                 "Show the logs of latest pipelinerun of given pipeline ",
+		Example:               eg,
+		Args:                  NameArg(p),
```
Comment (sthaha) [inline pkg/cmd/pipeline/logs.go]: Could you please move this validation to `validation` package?

File: pkg/cmd/pipeline/logs.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.PipelineName = args[0]
+			opts.Stream = &cli.Stream{
+				Out: cmd.OutOrStdout(),
+				Err: cmd.OutOrStderr(),
+			}
+			opts.Streamer = pods.NewStream
```
Comment (sthaha) [inline pkg/cmd/pipeline/logs.go]: Perhaps this should be default in `LogOptions` .. 
e.g 
```go
func (lo *LogOptions) Run() {
  streamer := pod.NewStream 
  if lo.Streamer != nil {
    streamer = lo.Streamer  
  }

 // use streamer here ... 
}

```

@hrishin wdyt?

Comment (hrishin) [reply pkg/cmd/pipeline/logs.go]: :+1: _works_

File: pkg/cmd/pipeline/logs.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		},
+	}
+	c.Flags().BoolVarP(&opts.AllSteps, "all", "a", false, "show all logs including init steps injected by tekton")
+	c.Flags().BoolVarP(&opts.Follow, "follow", "f", false, "stream live logs")
+	c.Flags().StringVar(&opts.PipelinerunName, "pipeline-run", "pr", "show logs for mentioned pipelinerun only")
+	return c
+}
+
+func latestRunName(cs *cli.Clients, ns string) (string, error) {
+	runs, err := cs.Tekton.TektonV1alpha1().PipelineRuns(ns).List(metav1.ListOptions{})
```
Comment (khrm) [inline pkg/cmd/pipeline/logs.go]: Is there any better method to get pipelineruns for a pipeline?

ATM, this code getting every pipelinerun from a namespace. Any upstream api?

@vdemeester @sthaha

File: pkg/cmd/pipeline/logs.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		return "", err
+	}
+
+	if len(runs.Items) == 0 {
+		return "", fmt.Errorf("No pipeline runs found in namespace: %s", ns)
+	}
+
+	var latest v1alpha1.PipelineRun
+	for _, run := range runs.Items {
+		if run.CreationTimestamp.After(latest.CreationTimestamp.Time) {
```
Comment (khrm) [inline pkg/cmd/pipeline/logs.go]: Shouldn't you check name also from spec?

File: pkg/cmd/pipeline/logs.go (hunk 8 above)
Comment (khrm) [reply pkg/cmd/pipeline/logs.go]: I just find out that we can use the label selector. Use that.

Comment (vdemeester) [reply pkg/cmd/pipeline/logs.go]: @khrm yes using labels you can filter. `latestRunName` needs to use `opts.PipelineName` as paramater.

```go
func latestRunName(cs *cli.Clients, pipelineName, ns string) {
// […]
}
```

File: pkg/cmd/pipeline/logs.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+				Out: cmd.OutOrStdout(),
+				Err: cmd.OutOrStderr(),
+			}
+			opts.Streamer = pods.NewStream
+
+			if opts.PipelineName == "" {
+				return fmt.Errorf("missing mandatory argument pipeline")
+			}
+
+			if opts.PipelinerunName == "" {
```
Comment (hrishin) [inline pkg/cmd/pipeline/logs.go]: Wondering why this check is required? :thinking:

File: pkg/cmd/pipeline/logs.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+				if err != nil {
+					return err
+				}
+			}
+			return opts.Run()
+		},
+	}
+	c.Flags().BoolVarP(&opts.AllSteps, "all", "a", false, "show all logs including init steps injected by tekton")
+	c.Flags().BoolVarP(&opts.Follow, "follow", "f", false, "stream live logs")
+	c.Flags().StringVar(&opts.PipelinerunName, "pipeline-run", "pr", "show logs for mentioned pipelinerun only")
```
Comment (hrishin) [inline pkg/cmd/pipeline/logs.go]: Wondering why `pipeline` command should accept `pipelinerun` as an argument? :thinking:, any specific reason?

As its already in `tkn pr logs <pr-name>`

File: pkg/cmd/pipeline/logs.go (hunk 8 above)
Comment (hrishin) [reply pkg/cmd/pipeline/logs.go]: Yes make sense if required at more than 2 places. :+1:

File: pkg/cmd/pipeline/pipeline.go [hunk 12]
```diff
@@ -35,6 +35,7 @@ func Command(p cli.Params) *cobra.Command {
 		describeCommand(p),
 		listCommand(p),
 		startCommand(p),
+		logCommand(p),
```
Comment (hrishin) [inline pkg/cmd/pipeline/pipeline.go]: could we maintain the alphabetical order? as we are doing it at other places as well

File: pkg/cmd/pipeline/logs.go (hunk 11 above)
Comment (sthaha) [reply pkg/cmd/pipeline/logs.go]: Agree with @hrishin , IMHO it just needs to ensure 

```
tkn p logs pipeline-name
```
shows the output of the `latest` pipelinerun for `pipeline-name`

File: pkg/cmd/pipeline/logs.go (hunk 4 above)
Comment (pradeepitm12) [reply pkg/cmd/pipeline/logs.go]: Done.

File: pkg/cmd/pipeline/logs.go (hunk 11 above)
Comment (pradeepitm12) [reply pkg/cmd/pipeline/logs.go]: This is to address the below requirement.
#57 
`There could also be flags to allow choosing pipelineruns prior to the latest`

File: pkg/cmd/pipeline/logs.go (hunk 8 above)
Comment (pradeepitm12) [reply pkg/cmd/pipeline/logs.go]: Yes,  it makes sense.

**Pull request merged**